central/cloudsources/manager/manager.go

```diff
@@ -138,18 +150,65 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
 		log.Errorw("Received errors during fetching assets from cloud sources. The result might be incomplete",
 			logging.Err(clientErrors))
 	}
-	log.Debugf("Got the following discovered clusters from Cloud Source integrations: %+v", discoveredClusters)
+	debugPrintDiscoveredClusters(discoveredClusters)
 	return discoveredClusters
 }
 
 func (m *managerImpl) reconcileDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
-	// TODO: Add matching of discovered clusters with secured clusters.
+	m.matchDiscoveredClusters(clusters)
+
+	debugPrintDiscoveredClusters(clusters)
 
 	if err := m.discoveredClustersDataStore.UpsertDiscoveredClusters(discoveredClusterCtx, clusters...); err != nil {
 		log.Errorw("Received errors during upserting discovered clusters.", logging.Err(err))
 	}
 }
 
+func (m *managerImpl) matchDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
+	// A list of hashes of currently secured cluster. A secured cluster hash consist of:
+	//  The secured cluster ID, and type.
+	securedClusters := set.NewStringSet()
+
+	unspecifiedClusterTypes := set.NewStringSet()
+
+	if err := m.clusterDataStore.WalkClusters(clustersCtx, func(obj *storage.Cluster) error {
+		providerMetadata := obj.GetStatus().GetProviderMetadata()
+
+		// Explicitly ignore clusters which do not have provider metadata associated with them.
+		if providerMetadata == nil {
+			return nil
+		}
+
+		// In case we have partial provider metadata (e.g. we miss the cluster name for EKS clusters, or the
+		// cluster type is unspecified), then mark the provider type as Unspecified.
+		// This means that we will assign each discovered cluster that cannot be safely matched to a secured cluster
+		// as Unspecified instead of Unsecured. This is to avoid false-positives.
+		clusterMetadata := providerMetadata.GetCluster()
+		if stringutils.AtLeastOneEmpty(clusterMetadata.GetId(), clusterMetadata.GetName()) {
```

Comment:
I'm not sure this does the correct thing now. Afaik we want the following behavior (for AWS specifically, but also generally):

Secured `/v1/clusters`:
* 1 EKS cluster 'A' with metadata ID and type correctly set.
* 1 EKS cluster 'B' with without metadata ID and type. This happens when instance tags are disabled, and results in `providerMetadata.Provider` being set, but `providerMetadata.ClusterMetadata` being nil.
* 1 OpenShift cluster 'C' being hosted on AWS with correct metadata.

Discovered clusters:
* 1 EKS cluster that matches 'A'. This should show as secured.
* 1 EKS cluster without match. This should show as unspecified.
* 1 OpenShift cluster on AWS (from OCM coming soon™) matching 'C'. This should show as secured.
* 1 OpenShift cluster on AWS without match. This should show as unsecured.

So in summary, EKS clusters in `/v1/clusters` may show up without Metadata type. So your idea with matching on provider level may be more correct after all.